File: web/test/staff-voices.test.js
import assert from 'node:assert/strict';
import test from 'node:test';
import { buildStaffVoices, splitStaffRests } from '../src/utils/staffVoices.js';
import { computeBeamGroups, slicePhraseIntoMeasures } from '../src/utils/sheetMusic.js';

const note = (pitch, startTime, duration) => ({ pitch, startTime, duration });

test('a held bass and moving melody retain independent durations and voices', () => {
    const notes = [note(48, 0, 2), note(72, 0, 0.5), note(74, 0.5, 0.5), note(76, 1, 0.5)];
    const voices = buildStaffVoices(notes);
    assert.equal(voices.length, 2);
    assert.deepEqual(voices[0].chords.map(chord => chord.notes.map(n => n.pitch)), [[72], [74], [76]]);
    assert.deepEqual(voices[1].chords.map(chord => chord.duration), [2]);
    assert.deepEqual(voices.map(voice => voice.rests.map(rest => [rest.startTime, rest.duration])), [[[1.5, 2.5]], [[2, 2]]]);
});

test('only exact simultaneous equal durations share stems, independently of display width', () => {
    const voices = buildStaffVoices([note(60, 0, 1), note(64, 0, 1), note(67, 0.0001, 1)]);
    assert.equal(voices.length, 2);
    assert.deepEqual(voices.flatMap(v => v.chords).map(c => c.notes.length), [1, 2]);
});

test('initial and internal silences are retained and an empty 6/8 staff gets a measure rest', () => {
    const [voice] = buildStaffVoices([note(60, 1, 0.5), note(62, 2, 1)], 0, 3);
    assert.deepEqual(voice.rests, [
        { startTime: 0, duration: 1, fullMeasure: false },
        { startTime: 1.5, duration: 0.5, fullMeasure: false },
    ]);
    assert.deepEqual(buildStaffVoices([], 6, 3)[0].rests, [{ startTime: 6, duration: 3, fullMeasure: true }]);
});

test('tied fragments preserve source identity and cover every voice without overlaps', () => {
    const phrase = { length: 2, tracks: { melody: [note(60, 3, 3), note(72, 4, 0.5)] } };
    const second = slicePhraseIntoMeasures(phrase)[1];
    const voices = buildStaffVoices(second.melodyNotes, second.measureStart, 4);
    const held = voices.flatMap(v => v.chords).flatMap(c => c.notes).find(n => n.pitch === 60);
    assert.equal(held.tieFromPrevious, true);
    assert.equal(held.duration, 2);
    assert.equal(phrase.tracks.melody[0].duration, 3);
    for (const voice of voices) {
        const spans = [...voice.chords, ...voice.rests].sort((a, b) => a.startTime - b.startTime);
        let cursor = 4;
        for (const span of spans) {
            assert.equal(span.startTime, cursor);
            cursor += span.duration;
        }
        assert.equal(cursor, 8);
    }
});

test('beams do not mix interleaved voices or simultaneous unequal durations', () => {
    const items = [
        { startBeat: 0, durationBeats: 0.25, flags: 2, voice: 0 },
        { startBeat: 0, durationBeats: 0.5, flags: 1, voice: 1 },
        { startBeat: 0.25, durationBeats: 0.25, flags: 2, voice: 0 },
        { startBeat: 0.5, durationBeats: 0.5, flags: 1, voice: 1 },
    ];
    assert.deepEqual(computeBeamGroups(items), [[0, 2], [1, 3]]);
    assert.deepEqual(computeBeamGroups(items.slice(0, 2).map((item) => ({
        startBeat: item.startBeat,
        durationBeats: item.durationBeats,
        flags: item.flags,
    }))), [[0], [1]]);
});

test('rest spelling respects compound pulses and preserves expressive residuals', () => {
    const span = { startTime: 0.5, duration: 2.5, fullMeasure: false };
    assert.deepEqual(splitStaffRests([span]).map(rest => rest.duration), [0.5, 1, 1]);
    assert.deepEqual(splitStaffRests([span], { numerator: 6, denominator: 8 }).map(rest => rest.duration), [1, 1.5]);
    const expressive = splitStaffRests([{ startTime: 0, duration: 0.26, fullMeasure: false }]);
    assert.equal(expressive.reduce((sum, rest) => sum + rest.duration, 0), 0.26);
    assert.equal(expressive.at(-1).notatable, false);
    assert.equal(splitStaffRests([{ startTime: 0, duration: 3, fullMeasure: true }]).length, 1);
});

test('a late upper voice is ranked above the held bass regardless of input order', () => {
    const notes = [note(48, 0, 4), note(72, 1, 1), note(74, 2, 1)];
    const voices = buildStaffVoices(notes);
    assert.deepEqual(voices[0].chords.flatMap(chord => chord.notes.map(n => n.pitch)), [72, 74]);
    assert.deepEqual(voices, buildStaffVoices(notes.slice().reverse()));
    assert.equal(voices[1].chords[0].notes[0].pitch, 48);
});

File: android/app/src/test/kotlin/com/tobietheunknown/pianoteacher/utils/StaffVoicesTest.kt
package com.tobietheunknown.pianoteacher.utils

import com.tobietheunknown.pianoteacher.data.model.NoteEvent
import org.junit.Assert.*
import org.junit.Test

class StaffVoicesTest {
    private fun note(pitch: Int, start: Double, duration: Double) =
        StaffDisplayNote(NoteEvent("$pitch-$start", pitch, start, duration), false, false)

    @Test fun heldBassAndMovingMelodyKeepIndependentDurations() {
        val voices = buildStaffVoices(listOf(note(48, 0.0, 2.0), note(72, 0.0, 0.5), note(74, 0.5, 0.5), note(76, 1.0, 0.5)), 4.0)
        assertEquals(2, voices.size)
        assertEquals(listOf(72, 74, 76), voices[0].chords.flatMap { it.notes }.map { it.note.pitch })
        assertEquals(listOf(2.0), voices[1].chords.map { it.duration })
        assertEquals(listOf(StaffRest(1.5, 2.5, false)), voices[0].rests)
        assertEquals(listOf(StaffRest(2.0, 2.0, false)), voices[1].rests)
    }

    @Test fun onlyEqualStartsAndDurationsShareStemsRegardlessOfZoom() {
        val voices = buildStaffVoices(listOf(note(60, 0.0, 1.0), note(64, 0.0, 1.0), note(67, 0.0001, 1.0)), 4.0)
        assertEquals(2, voices.size)
        assertEquals(listOf(1, 2), voices.flatMap { it.chords }.map { it.notes.size })
    }

    @Test fun restsFillInitialAndInternalGapsAndAnEmptyCompoundMeasure() {
        val voice = buildStaffVoices(listOf(note(60, 1.0, 0.5), note(62, 2.0, 1.0)), 3.0).single()
        assertEquals(listOf(StaffRest(0.0, 1.0, false), StaffRest(1.5, 0.5, false)), voice.rests)
        assertEquals(listOf(StaffRest(0.0, 3.0, true)), buildStaffVoices(emptyList(), 3.0).single().rests)
    }

    @Test fun tiedFragmentsRetainTiesAndEveryVoiceCoversTheMeasure() {
        val source = listOf(NoteEvent("held", 60, 3.0, 3.0), NoteEvent("moving", 72, 4.0, 0.5))
        val voices = buildStaffVoices(sliceNotesForStaff(source, 4.0, 8.0), 4.0)
        val held = voices.flatMap { it.chords }.flatMap { it.notes }.first { it.note.pitch == 60 }
        assertTrue(held.tieFromPrevious)
        assertEquals(2.0, held.note.duration, 0.0)
        assertEquals(3.0, source.first().duration, 0.0)
        for (voice in voices) {
            val spans = (voice.chords.map { it.startTime to it.duration } + voice.rests.map { it.startTime to it.duration }).sortedBy { it.first }
            var cursor = 0.0
            for ((start, duration) in spans) {
                assertEquals(cursor, start, 0.0)
                cursor += duration
            }
            assertEquals(4.0, cursor, 0.0)
        }
    }

    @Test fun beamsNeverMixInterleavedVoicesOrSimultaneousDurations() {
        val items = listOf(BeamItem(0.0, 0.25, 2, 0), BeamItem(0.0, 0.5, 1, 1), BeamItem(0.25, 0.25, 2, 0), BeamItem(0.5, 0.5, 1, 1))
        assertEquals(listOf(listOf(0, 2), listOf(1, 3)), computeBeamGroups(items))
        assertEquals(listOf(listOf(0), listOf(1)), computeBeamGroups(items.take(2).map { it.copy(voice = 0) }))
    }

    @Test fun restSpellingRespectsPulsesWithoutRoundingExpressiveTiming() {
        val rest = listOf(StaffRest(0.5, 2.5, false))
        assertEquals(listOf(0.5, 1.0, 1.0), splitStaffRests(rest).map { it.duration })
        assertEquals(listOf(1.0, 1.5), splitStaffRests(rest, 6, 8).map { it.duration })
        val expressive = splitStaffRests(listOf(StaffRest(0.0, 0.26, false)))
        assertEquals(0.26, expressive.sumOf { it.duration }, 1e-9)
        assertFalse(expressive.last().notatable)
        assertEquals(1, splitStaffRests(listOf(StaffRest(0.0, 3.0, true))).size)
    }

    @Test fun lateUpperVoiceIsRankedAboveBassRegardlessOfInputOrder() {
        val notes = listOf(note(48, 0.0, 4.0), note(72, 1.0, 1.0), note(74, 2.0, 1.0))
        val voices = buildStaffVoices(notes, 4.0)
        assertEquals(listOf(72, 74), voices[0].chords.flatMap { it.notes }.map { it.note.pitch })
        assertEquals(voices, buildStaffVoices(notes.reversed(), 4.0))
        assertEquals(48, voices[1].chords.single().notes.single().note.pitch)
    }
}
